plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -834,6 +844,120 @@ public void rollback()
+                    Optional.of(projectedColumns));
+
+            return Optional.of(new ProjectionApplicationResult<>(
+                    newTableHandle,
+                    projections,
+                    assignmentsList));
+        }
+
+        Map<String, Assignment> newAssignments = new HashMap<>();
+        ImmutableMap.Builder<ConnectorExpression, Variable> newVariablesBuilder = ImmutableMap.builder();
+        ImmutableSet.Builder<ColumnHandle> projectedColumnsBuilder = ImmutableSet.builder();
+
+        for (Map.Entry<ConnectorExpression, HiveApplyProjectionUtil.ProjectedColumnRepresentation> entry : columnProjections.entrySet()) {
+            ConnectorExpression expression = entry.getKey();
+            HiveApplyProjectionUtil.ProjectedColumnRepresentation projectedColumn = entry.getValue();
+
+            IcebergColumnHandle projectedColumnHandle;
+            String projectedColumnName;
+
+            // See if input already contains a columnhandle for this projected column, avoid creating duplicates.
+            // TODO: The Hive version finds an existing projection first, if it exists
+            // Optional<String> existingColumn = find(assignments, projectedColumn);
+
+            // Create a new column handle
+            IcebergColumnHandle baseColumnHandle = (IcebergColumnHandle) assignments.get(projectedColumn.getVariable().getName());
+            projectedColumnHandle = createProjectedColumnHandle(baseColumnHandle, projectedColumn.getDereferenceIndices(), expression.getType());
+            projectedColumnName = projectedColumnHandle.getName();
+
+            Variable projectedColumnVariable = new Variable(projectedColumnName, expression.getType());
+            Assignment newAssignment = new Assignment(projectedColumnName, projectedColumnHandle, expression.getType());
+            newAssignments.put(projectedColumnName, newAssignment);
+
+            newVariablesBuilder.put(expression, projectedColumnVariable);
+            projectedColumnsBuilder.add(projectedColumnHandle);
+        }
+
+        // Modify projections to refer to new variables
+        Map<ConnectorExpression, Variable> newVariables = newVariablesBuilder.build();
+        List<ConnectorExpression> newProjections = projections.stream()
+                .map(expression -> replaceWithNewVariables(expression, newVariables))
+                .collect(toImmutableList());
+
+        List<Assignment> outputAssignments = newAssignments.values().stream().collect(toImmutableList());
+        IcebergTableHandle newTableHandle = new IcebergTableHandle(
+                icebergTableHandle.getSchemaName(),
+                icebergTableHandle.getTableName(),
+                icebergTableHandle.getTableType(),
+                icebergTableHandle.getSnapshotId(),
+                icebergTableHandle.getUnenforcedPredicate(),
+                icebergTableHandle.getEnforcedPredicate(),
+                Optional.of(projectedColumnsBuilder.build()));
```

Comment:
That was based on the Hive implementation. Seems like the only place it's used is to avoid duplicating work if the same projections are given to the Metadata multiple times? https://github.com/trinodb/trino/pull/7360

It's also used for testing, but maybe we can skip it

Reply:
> Seems like the only place it's used is to avoid duplicating work if the same projections are given to the Metadata multiple times

This is actually important. The contract is that if call to `applyProjection` is a no-op should return Optional.empty()